core/src/main/scala/kafka/server/FetcherEventBus.scala

```diff
@@ -0,0 +1,164 @@
+    }
+  }
+
+  /**
+   * poll() unconditionally removes the earliest item
+   * If there are no items, poll() has no effect.
+   */
+  def poll(): Unit = {
+    delayedQueue.poll()
+  }
+
+  def size = delayedQueue.size
+}
+
+/**
+ * The FetcherEventBus supports queued events and delayed events.
+ * Queued events are inserted via the {@link #put} method, and delayed events
+ * are inserted via the {@link #schedule} method.
+ * Events are polled via the {@link #getNextEvent} method, which returns
+ * either a queued event or a scheduled event.
+ * @param time
+ */
+class FetcherEventBus(time: Time) {
+  private val eventLock = new ReentrantLock()
+  private val newEventCondition = eventLock.newCondition()
+
+  private val queue = new PriorityQueue[QueuedFetcherEvent]
+  private val scheduler = new SimpleScheduler[DelayedFetcherEvent]
+  @volatile private var shutdownInitialized = false
+
+  def eventQueueSize() = queue.size
+
+  def scheduledEventQueueSize() = scheduler.size
+
+  /**
+   * close should be called in a thread different from the one calling getNextEvent()
+   */
+  def close(): Unit = {
+    shutdownInitialized = true
+    inLock(eventLock) {
+      newEventCondition.signalAll()
+    }
+  }
+
+  def put(event: FetcherEvent): Unit = {
+    inLock(eventLock) {
+      queue.add(new QueuedFetcherEvent(event, time.milliseconds()))
+      newEventCondition.signalAll()
+    }
+  }
+
+  def schedule(delayedEvent: DelayedFetcherEvent): Unit = {
+    inLock(eventLock) {
+      scheduler.schedule(delayedEvent)
+      newEventCondition.signalAll()
+    }
+  }
+
+  /**
+   * There are 3 cases when the getNextEvent() method is called
+   * 1. There is at least one delayed event that has become current. If so, we return the delayed event with the earliest
+   * due time.
+   * 2. There is at least one queued event. If so, we return the queued event with the highest priority.
+   * 3. There are neither delayed events that have become current, nor queued events. We block until the earliest delayed
+   * event becomes current. A special case is that there are no delayed events at all, under which the call would block
+   * indefinitely until it is waken up by a new delayed or queued event.
+   *
+   * @return Either a QueuedFetcherEvent or a DelayedFetcherEvent that has become current. A special case is that the
+   *         FetcherEventBus is shutdown before an event can be polled, under which null will be returned.
+   */
+  def getNextEvent(): Either[QueuedFetcherEvent, DelayedFetcherEvent] = {
+    inLock(eventLock) {
+      var result : Either[QueuedFetcherEvent, DelayedFetcherEvent] = null
+
+      breakable {
+        while (!shutdownInitialized) {
+          val (delayedFetcherEvent, delayMs) = scheduler.peek()
+          if (delayedFetcherEvent.nonEmpty) {
+            scheduler.poll()
+            result = Right(delayedFetcherEvent.get)
+            break
+          } else if (!queue.isEmpty) {
+            result = Left(queue.poll())
```

Comment:
is it possible to get into a pathalogical situation here where we are constantly dequeuing delayed items that have their timer expired on every iteration, and never dequeue from `queue`?

It kinda feels like we should take the result of `scheduler.peek()` and put it into `queue`, and then only return items that come from the head of `queue`.

Reply:
That's a good point, I've updated the PR.